### PR TITLE
[Bug] Fix Wimp Out/EE Trainer Battle causing enemy stack with multi hit

### PR DESCRIPTION
**Pull request by muscode13**
## What are the changes the user will see?
Activating Wimp Out or Emergency Exit on a boss pokemon (Guzma 115) will no longer cause the ability to trigger twice and cause 2 pokemon to stack on top of one another

## Why am I making these changes?
Bug Thread: https://discord.com/channels/1125469663833370665/1290778513363959941

## How to test the changes?
Basically this issue started happening due to Wimp Out and Emergency Exit triggering twice due to the interaction with the boss break bars. One for Multi-hit, which waits until the end of damage, and the other is just for general damage. 

To prevent this I've added more guardrails around the applyAbAttrs to prevent this from happening again. Now there should be no overlap between the two applyAbAttrs.

I added an optional argument `move` to `damageAndUpdate` in `pokemon.ts`. This allows me to check for `MultiHitAttr`

I moved the multi-hit check to `move-effect-phase.ts` to utilize the `lastHit` variable. This will make the effect more consistent rather than relying on hitCount or hitsLeft.

## Before
<details>
  <summary>Multi-Hit Move</summary>

https://github.com/user-attachments/assets/c717072f-c72f-4bd3-8e50-aa626ed1bf58

</details>
<details>
  <summary>Multi-Lens</summary>

https://github.com/user-attachments/assets/963d77db-20a9-4421-8c62-6b89e61ea81e

</details>
<details>
  <summary>Parental Bond</summary>

https://github.com/user-attachments/assets/f7bd501f-b651-44c1-936a-f3dfcd949fb8

</details>

## After
<details>
  <summary>Multi-Hit Move</summary>

https://github.com/user-attachments/assets/53ca41e9-6029-46ed-9149-4392ec48013c

</details>
<details>
  <summary>Multi-Lens</summary>

https://github.com/user-attachments/assets/ec04cba4-c871-4dff-9cc1-3d5d51759fe2

</details>
<details>
  <summary>Parental Bond</summary>

https://github.com/user-attachments/assets/19f770c6-21ff-44bf-993c-997af4547f87

</details>

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`npm run test`)
  - [ ] Have I created new automated tests (`npm run create-test`) or updated existing tests related to the PR's changes?
- [x] Have I provided screenshots/videos of the changes (if applicable)?
  - [ ] Have I made sure that any UI change works for both UI themes (default and legacy)?

Are there any localization additions or changes? If so:
- [ ] Has a locales PR been created on the [locales](https://github.com/pagefaultgames/pokerogue-locales) repo?
  - [ ] If so, please leave a link to it here: 
- [ ] Has the translation team been contacted for proofreading/translation?

### CODE REVIEW

File: src/field/pokemon.ts [hunk 1]
```diff
@@ -3004,13 +2996,20 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
    * @param ignoreFaintPhase boolean to ignore adding a FaintPhase, passsed to damage()
    * @returns integer of damage done
    */
-  damageAndUpdate(damage: integer, result?: DamageResult, critical: boolean = false, ignoreSegments: boolean = false, preventEndure: boolean = false, ignoreFaintPhase: boolean = false, source?: Pokemon): integer {
+  damageAndUpdate(damage: integer, result?: DamageResult, critical: boolean = false, ignoreSegments: boolean = false, preventEndure: boolean = false, ignoreFaintPhase: boolean = false, source?: Pokemon, move?: Move): integer {
```
Comment (DayKev) [inline src/field/pokemon.ts:2999]:
```suggestion
  damageAndUpdate(damage: number, result?: DamageResult, critical: boolean = false, ignoreSegments: boolean = false, preventEndure: boolean = false, ignoreFaintPhase: boolean = false, source?: Pokemon, move?: Move): number {
```

**Review by DayKev — APPROVED** [review]

**Review by frutescens — APPROVED** [review]

File: src/phases/move-effect-phase.ts [hunk 2]
```diff
@@ -299,6 +302,16 @@ export class MoveEffectPhase extends PokemonPhase {
...
+            /**
+             * Multi-Lens, Multi Hit move and Parental Bond check for PostDamageAbAttr
+             * other damage source are calculated in damageAndUpdate in pokemon.ts
+             */
+            if (target.hasAbilityWithAttr(PostDamageForceSwitchAbAttr)) {
+              const multiHitModifier = user.getHeldItems().find(m => m instanceof PokemonMultiHitModifier);
+              if (multiHitModifier || user.hasAbilityWithAttr(AddSecondStrikeAbAttr) || move.hasAttr(MultiHitAttr)) {
+                applyPostDamageAbAttrs(PostDamageAbAttr, target, 0, target.hasPassive(), false, [], user);
+              }
+            }
```
Comment (innerthunder) [inline src/phases/move-effect-phase.ts:314]:
```suggestion
            if (user.turnData.hitCount > 1) {
              applyPostDamageAbAttrs(PostDamageAbAttr, target, 0, target.hasPassive(), false, [], user);
            }
```
Should be equivalent for a few reasons:
- `hitCount` should have been modified by all the different multi-hit modifiers at this point, either by this `MoveEffectPhase` or a previous `MoveEffectPhase` for the same action.
- `PostDamageAbAttrs` will only apply if the target actually has a post-damage attribute (that isn't suppressed/ignored), and EE is currently the only ability that has as much. This also makes post-damage attributes a bit more scalable.

You might have to fix the imports in this file with this change.

File: src/field/pokemon.ts [hunk 3]
```diff
@@ -3004,13 +2996,20 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
...
     damagePhase.updateAmount(damage);
-    applyPostDamageAbAttrs(PostDamageAbAttr, this, damage, this.hasPassive(), false, [], source);
+    /**
+     * Run PostDamageAbAttr from any source of damage that is not from a multi-hit
+     * Multi-hits are handled in move-effect-phase.ts for PostDamageAbAttr
+     */
+    const multiHitModifier = source?.getHeldItems().find(m => m instanceof PokemonMultiHitModifier);
+    if (!multiHitModifier && !source?.hasAbilityWithAttr(AddSecondStrikeAbAttr) && !move?.hasAttr(MultiHitAttr)) {
+      applyPostDamageAbAttrs(PostDamageAbAttr, this, damage, this.hasPassive(), false, [], source);
+    }
```
Comment (innerthunder) [inline src/field/pokemon.ts:3012]:
```suggestion
    if (source.turnData.hitCount <= 1) {
      applyPostDamageAbAttrs(PostDamageAbAttr, this, damage, this.hasPassive(), false, [], source);
    }
```
Similar deal here, whether or not multi-hit modifiers applied can be derived from `hitCount` at this point.

File: src/field/pokemon.ts [hunk 4]
```diff
@@ -3004,13 +2996,20 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
    * @param ignoreFaintPhase boolean to ignore adding a FaintPhase, passsed to damage()
    * @returns integer of damage done
    */
-  damageAndUpdate(damage: integer, result?: DamageResult, critical: boolean = false, ignoreSegments: boolean = false, preventEndure: boolean = false, ignoreFaintPhase: boolean = false, source?: Pokemon): integer {
+  damageAndUpdate(damage: number, result?: DamageResult, critical: boolean = false, ignoreSegments: boolean = false, preventEndure: boolean = false, ignoreFaintPhase: boolean = false, source?: Pokemon, move?: Move): number {
```
Comment (innerthunder) [inline src/field/pokemon.ts:2999]:
```suggestion
  damageAndUpdate(damage: number, result?: DamageResult, critical: boolean = false, ignoreSegments: boolean = false, preventEndure: boolean = false, ignoreFaintPhase: boolean = false, source?: Pokemon): number {
```
The other change I suggested here removes the need for the `move` arg.

**Comment by DayKev** [comment]
@innerthunder tested your version but it fails if there is no pokemon source for the damage (ie: stealth rocks). See: [Wimp Out test failures](https://github.com/pagefaultgames/pokerogue/actions/runs/12091042007/job/33718944840?pr=4935).

**Comment by DayKev** [comment]
Video of new method:

https://github.com/user-attachments/assets/411a99c0-2be0-49fc-86ce-5d4584759a1b

**Review by DayKev — APPROVED** [review]

**Review by MokaStitcher — APPROVED** [review]

**Review by frutescens — APPROVED** [review]

**Pull request merged**